File: nodejs/base/api/lib/routing/defaults/routes/default-delete-builder.ts
import { McmaResource, getTableName } from "@mcma/core";
import { DocumentDatabaseTableProvider } from "@mcma/data";

import { DefaultRouteHandlerConfigurator } from "../default-route-handler-configurator";
import { DefaultRouteBuilder } from "../default-route-builder";

export function defaultDeleteBuilder<T extends McmaResource>(
    dbTableProvider: DocumentDatabaseTableProvider,
    root: string
): DefaultRouteBuilder<T> {
    return new DefaultRouteBuilder<T>(
        "DELETE",
        root + "/{id}",
        new DefaultRouteHandlerConfigurator<T>((onStarted, onCompleted) =>
            async (requestContext) => {
                if (onStarted) {
                    const continueRequest = await onStarted(requestContext);
                    if (continueRequest !== undefined && !continueRequest) {
                        return;
                    }
                }
                const table = await dbTableProvider.get(getTableName(requestContext));
                
                const resource = await table.get<T>(requestContext.request.path);
                if (!resource) {
                    requestContext.setResponseResourceNotFound();
                    return;
                }
                await table.delete(requestContext.request.path);
                if (onCompleted) {
                    await onCompleted(requestContext, resource);
                }
            }
        )
    );
}

File: dotnet/base/Mcma.Api/Routing/Defaults/Routes/DefaultDeleteBuilder.cs
using System.Net.Http;
using Mcma.Context;
using Mcma.Data;

namespace Mcma.Api.Routing.Defaults.Routes
{
    public static class DefaultDeleteBuilder
    {
        public static DefaultRouteBuilder<TResource> Get<TResource>(IDocumentDatabaseTableProvider dbTableProvider, string root) where TResource : McmaResource
            => 
            new DefaultRouteBuilder<TResource>(
                HttpMethod.Delete,
                root + "/{id}",
                new DefaultRouteHandlerConfigurator<TResource>(
                    (onStarted, onCompleted) =>
                        async requestContext =>
                        {
                            // invoke the start handler, if any
                            if (onStarted != null)
                                if (!await onStarted(requestContext))
                                    return;

                            // get the table for the resource
                            var table = await dbTableProvider.GetAsync(requestContext.TableName());

                            // build id from the root public url and the path
                            var id = requestContext.CurrentRequestPublicUrl();

                            // get the resource from the db
                            var resource = await table.GetAsync<TResource>(id);

                            // if the resource doesn't exist, return a 404
                            if (resource == null)
                            {
                                requestContext.SetResponseResourceNotFound();
                                return;
                            }

                            // delete the resource from the db
                            await table.DeleteAsync(id);

                            // invoke the completion handler, if any
                            if (onCompleted != null)
                                await onCompleted(requestContext, resource);
                        }));
    }
}